src/v/lsm/db/memtable.cc

```diff
@@ -0,0 +1,194 @@
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "lsm/db/memtable.h"
+
+#include "absl/container/btree_map.h"
+#include "base/vassert.h"
+#include "lsm/core/internal/keys.h"
+#include "lsm/core/lookup_result.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <memory>
+#include <variant>
+
+namespace lsm::db {
+
+class memtable::iterator : public internal::iterator {
+public:
+    // The dummy iterator is only a place holder for the linked list in the
+    // memtable.
+    struct dummy {};
+    explicit iterator(dummy) {}
+
+    explicit iterator(ss::lw_shared_ptr<memtable> memtable)
+      : _mem(std::move(memtable))
+      , _it(_mem->_table.end()) {}
+    iterator(const iterator&) = delete;
+    iterator(iterator&&) = delete;
+    iterator& operator=(const iterator&) = delete;
+    iterator& operator=(iterator&&) = delete;
+    ~iterator() override {
+        if (_prev) {
+            _prev->_next = _next;
+        }
+        if (_next) {
+            _next->_prev = _prev;
+        }
+    }
+
+    bool valid() const override {
+        return ss::visit(
+          _it,
+          [](std::monostate) { return false; },
+          [this](memtable::table::iterator it) {
+              return it != _mem->_table.end();
+          },
+          [](const internal::key&) { return true; });
+    }
+
+    ss::future<> seek_to_first() override {
+        _it = _mem->_table.begin();
+        return ss::now();
+    }
+
+    ss::future<> seek_to_last() override {
+        _it = _mem->_table.empty() ? _mem->_table.end()
+                                   : std::prev(_mem->_table.end());
+        return ss::now();
+    }
+
+    ss::future<> seek(lsm::internal::key_view target) override {
+        _it = _mem->_table.lower_bound(target);
+        return ss::now();
+    }
+
+    ss::future<> next() override {
+        auto& it = restore();
+        if (it != _mem->_table.end()) {
+            ++it;
+        }
+        return ss::now();
+    }
+
+    ss::future<> prev() override {
+        auto& it = restore();
+        if (it == _mem->_table.begin()) {
+            it = _mem->_table.end();
+        } else if (it != _mem->_table.end()) {
+            --it;
+        } else if (!_mem->_table.empty()) {
+            it = std::prev(_mem->_table.end());
+        }
+        return ss::now();
+    }
+
+    lsm::internal::key_view key() override { return restore()->first; }
+
+    iobuf value() override { return restore()->second.share(); }
+
+    void invalidate() {
+        ss::visit(
+          _it,
+          [](std::monostate) {},
+          [this](memtable::table::iterator it) {
+              if (it == _mem->_table.end()) {
+                  _it = std::monostate{};
+              } else {
+                  _it = it->first;
+              }
```

Comment:
why isn't `_it` unconditionally set to monostate here (or rather, why the else case)? maybe i'm making assumptions about what invalidate means.